### PR TITLE
feat: add HTTP request headers to the dynamic multiple choice question

### PR DESCRIPTION
**Pull request by enthusiastio**
## Description
The dynamic multiple choice question is an important feature of the User Office Questionnaire. These questions allow for the real-time loading of answer sets from external sources. This pull request (PR) introduces the ability to specify HTTP headers for requests to external APIs when fetching answer sets

Changes
- HTTP Header Customization: This PR provides users with the option to define and customize HTTP headers included in requests to external APIs

- Issue Resolution: Additionally, this PR addresses an issue involving error logs during end-to-end integration tests. To mitigate this issue, a simple solution has been implemented. The system now checks for the presence of the URL before proceeding with the fetch operation. If the URL is absent, the fetch operation is aborted, eliminating the generation of error logs.

## Motivation and Context

Some APIs require authorization

## How Has This Been Tested

Added end-to-end tests

## Fixes

SWAP-3354

## Tests included/Docs Updated?

- [x] I have added tests to cover my changes.

image

### CODE REVIEW

**Review by ACLay — CHANGES_REQUESTED** [review]
Being able to set headers for these external api calls seems sensible, but I don't think we should be allowing the header values to be returned in plaintext via graphql. Feels like it'd be very easy for a misconfiguration of queries or  permissions to leak our credentials.

File: apps/backend/db_patches/db_seeds/0005_Questions.sql [hunk 1]
```diff
@@ -370,7 +370,7 @@ INSERT INTO templates_has_questions(
 VALUES 
   (
     'dynamic_multiple_choice_question', 
-    1, 5, 6, '{"variant":"dropdown","options":["One","Two","Three"],"isMultipleSelect":true}'
+    1, 5, 6, '{"variant":"dropdown", "url":"", "jsonPath":"","isMultipleSelect":true, "apiCallRequestHeaders":[]}'
```
Comment (bolmsten) [inline apps/backend/db_patches/db_seeds/0005_Questions.sql:373]: This patch will not run on an already existing db, we should create a new db patch

**Review by janosbabik — CHANGES_REQUESTED** [review]
Nice new feature; however, there are some changes that we need to make:
- Create a new database patch instead of modifying the old one.
- Protect the apiCallRequestHeaders field to prevent the leakage of any sensitive information.

File: apps/backend/src/resolvers/types/FieldConfig.ts [hunk 2]
```diff
@@ -104,8 +113,8 @@ export class DynamicMultipleChoiceConfig extends ConfigBase {
   @Field(() => [String])
   options: string[];
 
-  @Field(() => Boolean)
-  externalApiCall: boolean;
+  @Field(() => [ApiCallRequestHeader])
```
[the remark this answers is not in the document]
Comment (enthusiastio) [reply apps/backend/src/resolvers/types/FieldConfig.ts:116]: Thanks @janosbabik, thanks for the suggestion to do it in the resolver. I was initally planning it to do somehwere else, but this is better :thumbsup:

**Comment by enthusiastio** [comment]
> Being able to set headers for these external api calls seems sensible, but I don't think we should be allowing the header values to be returned in plaintext via graphql. Feels like it'd be very easy for a misconfiguration of queries or permissions to leak our credentials.

Thanks @ACLay for the early catch :eagle:

File: apps/backend/db_patches/db_seeds/0005_Questions.sql (hunk 1 above)
Comment (Junjiequan) [reply apps/backend/db_patches/db_seeds/0005_Questions.sql:373]: If I'm not mistaken, all the patches in the db_seeds folder serve E2E test only

File: apps/backend/src/resolvers/types/FieldConfig.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (janosbabik) [reply apps/backend/src/resolvers/types/FieldConfig.ts:127]: @enthusiastio: I did not notice that the sql file modification was in the db_seeds directory, sorry for that :)

**Review by ACLay — APPROVED** [review]

**Pull request merged**